Source/options.cpp

```diff
@@ -572,6 +568,90 @@ std::vector<OptionEntryBase *> AudioOptions::GetEntries()
 	};
 }
 
+OptionEntryResolution::OptionEntryResolution()
+    : OptionEntryListBase("", OptionEntryFlags::CantChangeInGame | OptionEntryFlags::RecreateUI, N_("Resolution"), N_("Affect the game's internal resolution and determine your view area. Note: This can differ from screen resolution, when Upscaling, Integer Scaling or Fit to Screen is used."))
+{
+}
+void OptionEntryResolution::LoadFromIni(string_view category)
+{
+	size = { GetIniInt(category.data(), "Width", DEFAULT_WIDTH), GetIniInt(category.data(), "Height", DEFAULT_HEIGHT) };
+}
+void OptionEntryResolution::SaveToIni(string_view category) const
+{
+	SetIniValue(category.data(), "Width", size.width);
+	SetIniValue(category.data(), "Height", size.height);
+}
+
+void OptionEntryResolution::CheckResolutionsAreInitialized() const
+{
+	if (!resolutions.empty())
+		return;
+
+	std::vector<Size> sizes;
+
+	// Add resolutions
+#ifdef USE_SDL1
+	auto *modes = SDL_ListModes(nullptr, SDL_FULLSCREEN | SDL_HWPALETTE);
+	if (modes != nullptr) {
+		for (size_t i = 0; modes[i] != nullptr; i++) {

```

Comment:
This crashes on 3DS. We're dealing with an.. *interesting* error code.

https://www.libsdl.org/release/SDL-1.2.15/docs/html/sdllistmodes.html
```c++
/* Check if our resolution is restricted */
if(modes == (SDL_Rect **)-1){
  printf("All resolutions available.\n");
}
```

Reply:
Very interesting.
I adjusted the code. Could you take a look again?

Reply:
This works. With this change, I can toggle between 800x480 and 640x480. Changing the resolution absolutely destroys the quality of the graphics for some reason, but that's probably only tangentially related to this PR, like the `PalSurface` issue. It looks like scaling artifacts.

![image](https://user-images.githubusercontent.com/9203145/145107753-285c342b-47ed-4760-9f3e-b8bd0d3a23b0.png)

Reply:
It might be performing the scaling early in software instead of late in hardware.